pkg/stub/handler.go

```diff
@@ -113,60 +116,67 @@ type Handler struct {
-			return nil
-		}
-		cm = &corev1.ConfigMap{}
-		cm.Name = v1alpha1.SamplesResourceName
-		cm.Data = map[string]string{}
-		if len(srcfg.Spec.InstallType) == 0 {
-			cm.Data[installtypekey] = string(v1alpha1.CentosSamplesDistribution)
-		} else {
-			cm.Data[installtypekey] = string(srcfg.Spec.InstallType)
-		}
-		if len(srcfg.Spec.Architectures) == 0 {
-			cm.Data[v1alpha1.X86Architecture] = v1alpha1.X86Architecture
-		} else {
-			for _, arch := range srcfg.Spec.Architectures {
-				switch arch {
-				case v1alpha1.X86Architecture:
-					cm.Data[v1alpha1.X86Architecture] = v1alpha1.X86Architecture
-				case v1alpha1.PPCArchitecture:
-					cm.Data[v1alpha1.PPCArchitecture] = v1alpha1.PPCArchitecture
-				}
+func (h *Handler) StoreCurrentValidConfig(condition v1alpha1.SamplesResourceConditionType, srcfg *v1alpha1.SamplesResource) error {
+
+	if condition != v1alpha1.SamplesExist {
+		return nil
+	}
+
+	cm, err := h.configmapclientwrapper.Get(h.namespace, v1alpha1.SamplesResourceName)
+	if err != nil && !kerrors.IsNotFound(err) {
+		// just return error to sdk for retry
+		return err
+	}
+	if kerrors.IsNotFound(err) {
+		// 4.0 testing showed that we were getting empty ConfigMaps
+		cm = nil
+	}
+	if cm != nil {
+		// just return ... we only update the config map once
+		return nil
+	}
+	cm = &corev1.ConfigMap{}
+	cm.Name = v1alpha1.SamplesResourceName
+	cm.Data = map[string]string{}
+	if len(srcfg.Spec.InstallType) == 0 {
+		cm.Data[installtypekey] = string(v1alpha1.CentosSamplesDistribution)
+	} else {
+		cm.Data[installtypekey] = string(srcfg.Spec.InstallType)
+	}
+	if len(srcfg.Spec.Architectures) == 0 {
+		cm.Data[v1alpha1.X86Architecture] = v1alpha1.X86Architecture
+	} else {
+		for _, arch := range srcfg.Spec.Architectures {
+			switch arch {
+			case v1alpha1.X86Architecture:
+				cm.Data[v1alpha1.X86Architecture] = v1alpha1.X86Architecture
+			case v1alpha1.PPCArchitecture:
+				cm.Data[v1alpha1.PPCArchitecture] = v1alpha1.PPCArchitecture
 			}
 		}
-		_, err = h.configmapclientwrapper.Create(h.namespace, cm)
-		if err != nil {
-			// just return error to sdk for retry
-			return err
-		}
 	}
-	return nil
+	_, err = h.configmapclientwrapper.Create(h.namespace, cm)
+	return err
 }
 
 func (h *Handler) SpecValidation(srcfg *v1alpha1.SamplesResource) error {
+	// TODO - the first thing this should do is check that all the config values
+	// are "valid" (the architecture name is known, the distribution name is known, etc)
+	// if that fails, we should immediately error out and set ConfigValid to false.
+
+	// only if the values being requested are valid, should we then proceed to check
+	// them against the previous values(if we've stored any previous values)
+
 	// if we have not had a valid SamplesResource processed, allow caller to try with
 	// the srcfg contents
-	if h.samplesResource == nil || !h.samplesResource.ConditionTrue(v1alpha1.SamplesExist) {
+	if !srcfg.ConditionTrue(v1alpha1.SamplesExist) {
```

Comment:
this looked suspect to me.  i'd like to get rid of as much of this "h.samplesResource" caching logic as possible as it's difficult to reason about the state of that value when the operator has restarted.

`srcfg` represents the current state of the system (the event we just got), so use that to drive decision making about whether we need to compare with an old config or not.

Reply:
> So either here, or in a follow up PR, I think we are good for changing this to be a string, and remove any fears of a cached, full copy.

definitely a follow up.  Not 100% sure i understand the remaining use case (not sure why it would be a string vs a boolean?) but we can discuss that in the followup that reworks that logic/removes the cache field.

Reply:
OK, the second item for my follow up PR.